### PR TITLE
fix: DESCRIBE does not show primary key

### LINKED ISSUE #1468: DESCRIBE does not show primary key
Version 0.2.4 on Linux

```
D CREATE TABLE t1(id INTEGER PRIMARY KEY, j VARCHAR);
D describe t1;
┌───────┬─────────┬──────┬─────┬─────────┬───────┐
│ Field │  Type   │ Null │ Key │ Default │ Extra │
├───────┼─────────┼──────┼─────┼─────────┼───────┤
│ id    │ INTEGER │ NO   │     │         │       │
│ j     │ VARCHAR │ YES  │     │         │       │
└───────┴─────────┴──────┴─────┴─────────┴───────┘
D PRAGMA table_info('t1');
┌─────┬──────┬─────────┬─────────┬────────────┬───────┐
│ cid │ name │  type   │ notnull │ dflt_value │  pk   │
├─────┼──────┼─────────┼─────────┼────────────┼───────┤
│ 0   │ id   │ INTEGER │ true    │            │ true  │
│ 1   │ j    │ VARCHAR │ false   │            │ false │
└─────┴──────┴─────────┴─────────┴────────────┴───────┘
```

### PR DESCRIPTION
**Pull request by gkaretka**
fixes #1468

### CODE REVIEW

**Comment by gkaretka** [comment]
#1468

**Comment by Tishj** [comment]
Thanks for the PR, could you add `fixes #1468` to the description of the PR? That way it will automatically close the assigned issue :)

File: src/function/table/system/pragma_table_info.cpp [hunk 1]
```diff
@@ -50,6 +50,9 @@ static unique_ptr<FunctionData> PragmaTableInfoBind(ClientContext &context, Tabl
 	names.emplace_back("pk");
 	return_types.emplace_back(LogicalType::BOOLEAN);
 
+	names.emplace_back("unique");
+	return_types.emplace_back(LogicalType::BOOLEAN);
```
Comment (Mytherin) [inline src/function/table/system/pragma_table_info.cpp:54]: `PRAGMA table_info` is specifically modelled after SQLite's equivalent `PRAGMA table_info` - so we don't want to add new columns here.

File: test/sql/pragma/test_show_tables.test [hunk 2]
```diff
@@ -11,19 +11,43 @@ CREATE TABLE "select"(i INTEGER);
...
+
+statement ok
+CREATE TABLE t2 (id INTEGER PRIMARY KEY, j VARCHAR UNIQUE)
+
+# column_name | column_type | null | key | default | extra
+query TTTITI
+DESCRIBE t1
+----
+id	INTEGER	NO	PRI	NULL	(empty)
+j	VARCHAR	YES	(empty)	NULL	(empty)
```
Comment (Mytherin) [inline test/sql/pragma/test_show_tables.test:25]: `DESCRIBE` is modeled after MySQL, here is what MySQL outputs for this specific table:

```
mysql> CREATE TABLE t2 (id INTEGER PRIMARY KEY, j VARCHAR(100) UNIQUE);
Query OK, 0 rows affected (0.01 sec)

mysql> DESCRIBE t2;
+-------+--------------+------+-----+---------+-------+
| Field | Type         | Null | Key | Default | Extra |
+-------+--------------+------+-----+---------+-------+
| id    | int          | NO   | PRI | NULL    |       |
| j     | varchar(100) | YES  | UNI | NULL    |       |
+-------+--------------+------+-----+---------+-------+
2 rows in set (0.00 sec)

```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR!

This is changing a bit too much in a non-backwards compatible manner. Instead of changing `PRAGMA table_info` - perhaps we can rewrite `PragmaShow` to use the other catalog functions instead. For example, we can use the `duckdb_constraints` and `duckdb_columns` functions which already have this information:

```sql
CREATE TABLE t2 (id INTEGER PRIMARY KEY, j VARCHAR UNIQUE);
SELECT 
	column_name,
	data_type AS "column_type",
	CASE WHEN is_nullable THEN 'YES' ELSE 'NO' END AS "null",
	(
		SELECT MIN(
			CASE WHEN constraint_type='PRIMARY KEY' THEN 'PRI'
			WHEN constraint_type='UNIQUE' THEN 'UNI'
			ELSE NULL END)
		FROM duckdb_constraints() c
		WHERE c.table_oid=cols.table_oid AND list_contains(constraint_column_names, cols.column_name)
	) AS "key",
	column_default AS "default",
	NULL AS "extra"
FROM duckdb_columns cols WHERE table_name='t2';
┌─────────────┬─────────────┬─────────┬─────────┬─────────┬───────┐
│ column_name │ column_type │  null   │   key   │ default │ extra │
│   varchar   │   varchar   │ varchar │ varchar │ varchar │ int32 │
├─────────────┼─────────────┼─────────┼─────────┼─────────┼───────┤
│ id          │ INTEGER     │ NO      │ PRI     │ NULL    │  NULL │
│ j           │ VARCHAR     │ YES     │ UNI     │ NULL    │  NULL │
└─────────────┴─────────────┴─────────┴─────────┴─────────┴───────┘
``` 

This way the other tests can be left as-is and we really only change the output of `SHOW` to correctly show unique constraints/primary keys.

File: test/sql/pragma/test_show_tables.test (hunk 2 above)
Comment (gkaretka) [reply test/sql/pragma/test_show_tables.test:25]: Thanks for comments I will undo some changes and push it here

File: src/function/pragma/pragma_queries.cpp [hunk 3]
```diff
@@ -58,10 +58,27 @@ string PragmaFunctionsQuery(ClientContext &context, const FunctionParameters &pa
...
-	return StringUtil::Format(
-	    "SELECT name AS \"column_name\", type as \"column_type\", CASE WHEN \"notnull\" THEN 'NO' ELSE 'YES' "
-	    "END AS \"null\", NULL AS \"key\", dflt_value AS \"default\", NULL AS \"extra\" FROM pragma_table_info('%s');",
-	    parameters.values[0].ToString());
+
+	// Remove leading and trailing quotes from string
+	auto table_name = parameters.values[0].ToString();
+	StringUtil::RemoveFirstAndLastIfMatching(table_name, '"');
+
+	return StringUtil::Format("SELECT name AS \"column_name\", "
```
Comment (Mytherin) [inline src/function/pragma/pragma_queries.cpp:66]: Can we use a multi-line string literal here, similar to what is used in the [FTS extension](https://github.com/duckdb/duckdb/blob/master/extension/fts/fts_indexing.cpp#L37)?

File: src/function/pragma/pragma_queries.cpp [hunk 4]
```diff
@@ -58,10 +58,27 @@ string PragmaFunctionsQuery(ClientContext &context, const FunctionParameters &pa
...
 string PragmaShow(ClientContext &context, const FunctionParameters &parameters) {
 	// PRAGMA table_info but with some aliases
-	return StringUtil::Format(
-	    "SELECT name AS \"column_name\", type as \"column_type\", CASE WHEN \"notnull\" THEN 'NO' ELSE 'YES' "
-	    "END AS \"null\", NULL AS \"key\", dflt_value AS \"default\", NULL AS \"extra\" FROM pragma_table_info('%s');",
-	    parameters.values[0].ToString());
+
+	// Remove leading and trailing quotes from string
+	auto table_name = parameters.values[0].ToString();
+	StringUtil::RemoveFirstAndLastIfMatching(table_name, '"');
```
Comment (Mytherin) [inline src/function/pragma/pragma_queries.cpp:64]: Can we use `QualifiedName::Parse` here instead? That is what `pragma_table_info` uses internally - and also allows us to support tables in different schemas, e.g.:

```sql
create schema s1;
create table s1.tbl(i int);
create table tbl(j int);
describe s1.tbl;
┌─────────────┬─────────────┬─────────┬───────┬─────────┬───────┐
│ column_name │ column_type │  null   │  key  │ default │ extra │
│   varchar   │   varchar   │ varchar │ int32 │ varchar │ int32 │
├─────────────┼─────────────┼─────────┼───────┼─────────┼───────┤
│ i           │ INTEGER     │ YES     │  NULL │ NULL    │  NULL │
└─────────────┴─────────────┴─────────┴───────┴─────────┴───────┘
describe tbl;
┌─────────────┬─────────────┬─────────┬───────┬─────────┬───────┐
│ column_name │ column_type │  null   │  key  │ default │ extra │
│   varchar   │   varchar   │ varchar │ int32 │ varchar │ int32 │
├─────────────┼─────────────┼─────────┼───────┼─────────┼───────┤
│ j           │ INTEGER     │ YES     │  NULL │ NULL    │  NULL │
└─────────────┴─────────────┴─────────┴───────┴─────────┴───────┘
```

Could we also add that as a test case?

File: src/common/string_util.cpp [hunk 5]
```diff
@@ -26,6 +26,17 @@ void StringUtil::LTrim(string &str) {
 	str.erase(str.begin(), it);
 }
 
+// Removes first and last char
+void StringUtil::RemoveFirstAndLast(string &str) {
+	str = str.substr(1, str.size() - 2);
+}
+
+void StringUtil::RemoveFirstAndLastIfMatching(string &str, char c) {
+	if (str.front() == c && str.back() == c) {
```
Comment (Mytherin) [inline src/common/string_util.cpp:35]: Could we remove these and their tests again? They are rather specific and not necessary in this case.

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates! Looks good - some more comments.

**Comment by Mytherin** [comment]
Also there is no need to continuously merge with master and push. It is re-triggering the CI unnecessarily and taking up CI time on the main repo.

**Comment by Mytherin** [comment]
Please stop merging with master to retrigger CI. I have asked you this before. It is consuming shared CI resources.

If there is an unrelated CI failure due to a network outage you can state that. You can also look at other PRs and see if they have the same problem.

**Comment by gkaretka** [comment]
> Please stop merging with master to retrigger CI. I have asked you this before. It is consuming shared CI resources.
> 
> If there is an unrelated CI failure due to a network outage you can state that. You can also look at other PRs and see if they have the same problem.

Sorry for that, I just merged most recent changes to my fork

**Comment by Mytherin** [comment]
Thanks for the updates - looks good to me

**Pull request merged**